theia-ide/theia
# upgrading to LSP 5.3.0 and Monaco 0.17.0

#### What it does


- upgrading to LSP 5.3.0
  - LSP dependencies are targeting es6 now, we are forced to use babel in order to transpile es6 to es5 classes for these dependencies
- upgrading to Monaco 0.17.0
  - [x] open a CQ for all copied code to align with latest vscode.d.ts - it was already covered by previous CQs
- [x] reference peek cannot be closed with `esc`
- [x] monaco configuration does not reflect editor preferences
- [x] fix editor preview tests
- [x] depends on https://github.com/TypeFox/monaco-languageclient/pull/174
- [x] https://github.com/theia-ide/theia/pull/5901#issuecomment-522919840
- [ ] https://github.com/theia-ide/theia/pull/5901#issuecomment-528796882
- [x] https://github.com/theia-ide/theia/pull/5901#issuecomment-529311130

#### How to test


- test language features with native Theia extension:
  - test that features are available in the editor context menu and in quick command palette
  - a declaration is revealed when following the reference
  - focus belongs to a proper editor after following a reference (for internal and outside references, from editor to editor preview and view versa)
  - reference peek is still opened after following a reference (for internal and outside references, from editor to editor preview and view versa)
  - esc closes reference peek widget
  - completion items are properly resolved and inserted
  - focus belongs to a proper editor on peek definition (for internal and outside references, from editor to editor preview and view versa)
  - check that rename work inside the file and across files
  - check that arguments are proposed by the signature help
  - check that hover is present with properly rendered docs
  - check that symbol occurrences is highlighted when a declaration or reference is selected
  - check that go to (type|implementation) definition works properly, i.e. definition is revealed
  - check that code lens work properly
    - run/debug code lenses: https://github.com/kasecato/vscode-javadebug-sample
  - check that code actions work properly (quick fixes)
  - check that document and selection formatting work properly
  - check color provider with css colors
  - check link provider with links in html
  - check that folding ranges are properly provided for js, css and html
  - check go to symbol command
  - check open workspace symbol command
- test language features with VS Code extension:
  - typesctipt VS Code extension for testing: [ts_plugins.zip](https://github.com/theia-ide/theia/files/3513239/ts_plugins.zip)
  - to run it:
    - uninstall native typescript extension in example package.json
    - build the example app
    - unzip archive into plugins folder
    - start the example app
  - use the same checks as for native Theia extension
- test quick palette:
  - proper keys are displayed for the quick command palette, i.e. they are displayed and a corresponding Monaco command can be triggered by such keybinding
  - use sample vscode extension
  - use fuzzy search
- test vscode tree view (that context keys work properly)
- test plugin editor decorations, for example with power mode extension
- test debug console input editor
- test debug breakpoint editor
- test that preferences are propagated as config key values: change editor preferences and check whether they are reflected
- test language auto detection
- test `edit` menu with editor
- test `selection` menu with editor
- test editor context menu
- smoke test some VS Code extensions like go, python, just use editor for a bit, check backend and frontend logs for errors

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
